### PR TITLE
fix(proxy): preserve client IPs behind a local reverse proxy

### PR DESCRIPTION
**Pull request by cymen**
When Temps runs behind nginx on the same host, the proxy sees the loopback TCP peer and previously replaced `X-Forwarded-For` with that address. Analytics, proxy logs, and IP-based controls then used the local proxy instead of the visitor.

This PR adds **opt-in** client-IP resolution for loopback peers. An admin can enable it under **Settings → Security → Proxy client IP**; the database-backed setting reaches both `temps serve` and standalone `temps proxy` processes within a few seconds, with no database I/O on the request path. Trust is off by default. This is the sole control surface — an earlier revision of this PR also added a `TEMPS_TRUST_LOOPBACK_FORWARDED_IP` / `--trust-loopback-forwarded-ip` startup override, but that duplicated an already DB-backed, admin-UI-controlled, audited setting and could silently keep trust enabled after an admin explicitly turned it off, which conflicts with this repository's rule against environment-variable-backed runtime configuration. It has been removed; use the Security page for every deployment topology, including a standalone `temps proxy` process.

Enable it only when the local reverse proxy overwrites `X-Real-IP` and either overwrites `X-Forwarded-For` with the observed client address or appends that address as the final entry. The installation guide includes an nginx example. Passing client-supplied values through unchanged permits IP spoofing.

The resolver accepts forwarding headers only from loopback peers, including IPv4-mapped IPv6 loopback addresses (`::ffff:127.0.0.1`) seen on dual-stack listeners. It uses the rightmost address in the last XFF field, falling back to XRI only when XFF is absent; missing, malformed, or non-IP values fall back to the TCP peer. Existing Cloudflare and Bunny CDN handling is preserved. An older settings client that omits the new field cannot silently reset the admin's choice. The saved setting is loaded before the proxy accepts traffic; an initial read failure aborts startup rather than guessing at the IP policy. Later failed refreshes retain the last successfully loaded value. The refresh task releases its strong snapshot reference before each polling sleep and exits when its proxy service is dropped. Changing the setting now emits a dedicated `FORWARDED_IP_TRUST_UPDATED` audit event with the previous and new value.

## Evidence

- `cargo test --lib -p temps-core -p temps-config -p temps-proxy`: **1,149 passed, 4 ignored**. Covers both settings PUT preservation and explicit disable, and proxy client-IP resolution. The earlier proxy/CLI suite also passed (**932 passed, 4 ignored**) after the flag and env-variable change.
- `cargo test --lib -p temps-proxy` on the current change: **628 passed, 4 ignored**. The startup tests assert that a saved admin choice is already in the initial atomic value and that a settings-read error stops startup. The test fixture uses an OS-assigned port after a local run hit an occupied hard-coded port 9000; it is aligned with the same fix and distinct-port test now on `main`. `cargo check --lib`: **0 errors**. `bunx tsc --noEmit` and scoped ESLint passed. The web SDK was regenerated from this worktree's running authenticated OpenAPI endpoint with `TEMPS_OPENAPI_URL=http://localhost:8180/api/api-docs/openapi.json bun run openapi-ts` (**5 files generated**, source attribution added).
- Live check in an isolated `/start-temps` slot (`8180` proxy, `3010` web, `temps_dev_s10` database). Created a synthetic `203.0.113.7/32` block rule through `POST /api/ip-access-control` (**201**). With the admin opt-in off, `curl -H 'Host: pr971.local.test' -H 'X-Forwarded-For: 203.0.113.7' http://127.0.0.1:8180/pr971-ip-check-off` returned **200**. In `agent-browser`, opened `http://localhost:3010/settings/security`, expanded **Proxy client IP**, saw the switch off, enabled it, clicked **Save Changes**, and saw **Security settings saved**. `GET /api/settings` returned `{"trust_loopback_forwarded_ip":true}`; the same curl on `/pr971-ip-check-on` returned **403** with `X-Blocked-Reason: IP address blocked`.
- Restarted `temps serve`. Startup logged `Loaded loopback forwarded-IP trust before proxy startup enabled=true` before the proxy listener was added. The same request on `/pr971-ip-check-restart` returned **403** with the block reason, and reloading the Security page showed the switch still on. This current-head loopback request exercises the opt-in path; the earlier beta.56 nginx run verified the reverse-proxy configuration and visitor attribution before the opt-in was added.
- Follow-up review pass (3 commits, `23e09f9`..`cd1b0de`): removed the CLI/env startup override for the CLAUDE.md reasons above, fixed the IPv4-mapped IPv6 loopback-detection gap, and added the `FORWARDED_IP_TRUST_UPDATED` audit event. `cargo test --lib -p temps-proxy -p temps-cli -p temps-config` after these commits: **1042 passed, 4 ignored**. `cargo check --lib`, `cargo fmt --check`, and `cargo clippy --all-targets -- -D warnings` (invoked directly, not through the token-optimizing proxy some earlier evidence runs used) are all clean on the affected crates.

The workspace emits existing CAPTCHA profile and `proc-macro-error2` future-compatibility notices. Existing visitor profiles retain their historical IPs; this change affects new request attribution and does not rewrite old records.

### CODE REVIEW

**Comment by dviejokfs** [comment]
Could we make loopback forwarded-header trust opt-in (off by default), ideally scoped to an explicitly trusted proxy connection or listener?

At `client_ip.rs:17-43`, any loopback TCP peer can supply the rightmost `X-Forwarded-For` value (or `X-Real-IP` when XFF is absent). A plain nginx `proxy_pass` forwards incoming request headers unless configured otherwise. In that setup, a remote client can send `X-Forwarded-For: <unblocked IP>` unchanged; the new resolver then uses that address for the proxy's IP block check and the XFF sent upstream. Existing local-proxy installations would gain this trust automatically on upgrade.

For the opt-in path, the nginx example should explicitly overwrite XFF with `$remote_addr` (or append the observed peer address) and overwrite X-Real-IP as well. Please test the default-off behavior and the configured path. The flag controls when Temps trusts the header; it cannot make an unsanitized header trustworthy by itself.

File: crates/temps-proxy/src/server.rs [hunk 1]
```diff
@@ -32,6 +34,44 @@ use async_trait::async_trait;
...
+        loop {
+            let enabled = match config_service.get_settings().await {
+                Ok(settings) => effective_forwarded_ip_trust(
+                    startup_enabled,
+                    settings.trust_loopback_forwarded_ip(),
+                ),
+                Err(error) => {
+                    tracing::warn!(%error, "Could not refresh forwarded-IP trust; using the safe startup default");
+                    startup_enabled
+                }
```
[the remark this answers is not in the document]
Comment (dviejokfs) [reply crates/temps-proxy/src/server.rs:88]: Fixed in e6015034. A failed settings refresh now retains the last successfully loaded trust value; an instance that has never loaded the setting still starts with trust off unless the startup override is enabled. Added a regression test covering enable, read failure, disable, and the startup override. The proxy suite passed (624 passed, 4 ignored), as did cargo check and Clippy.

File: crates/temps-proxy/src/server.rs [hunk 2]
```diff
@@ -32,6 +34,52 @@ use async_trait::async_trait;
...
+}
+
+fn start_forwarded_ip_trust_refresh(
+    config_service: Arc<temps_config::ConfigService>,
+    startup_enabled: bool,
+) -> Result<Arc<AtomicBool>> {
+    let effective = Arc::new(AtomicBool::new(startup_enabled));
+    let snapshot = Arc::downgrade(&effective);
+    security_refresh_runtime()?.spawn(async move {
+        while let Some(snapshot) = snapshot.upgrade() {
```
[the remark this answers is not in the document]
Comment (dviejokfs) [reply crates/temps-proxy/src/server.rs:82]: Addressed in 8e11f01. The loop-bound strong reference was released at the end of each iteration, so it did not persist forever, but it was held across the polling sleep. The task now explicitly drops that reference before sleeping; after the LoadBalancer is dropped, the next weak upgrade fails and the task exits before another settings read. The full proxy suite passed (624 passed, 4 ignored), and the repository hooks passed.

**Comment by dviejokfs** [comment]
@cymen I pushed a few follow-ups to this PR so your client-IP fix works safely behind a local reverse proxy:

- Made loopback forwarded-header trust **off by default**, with `TEMPS_TRUST_LOOPBACK_FORWARDED_IP=true` / `--trust-loopback-forwarded-ip` opt-ins. A plain nginx proxy can pass client-supplied forwarding headers through, so enabling trust automatically would let visitors spoof the IP used by analytics and IP-based controls.
- Added an admin switch at **Settings → Security → Proxy client IP**. Proxy processes pick up the saved setting at runtime; older settings clients cannot silently reset it. The startup opt-in still forces trust on for that process, so operators retain an explicit configuration override.
- Kept the last successfully loaded trust setting when a database refresh fails, avoiding a temporary return to loopback attribution. The refresh task now releases its strong snapshot reference before sleeping, so it stops after its proxy service is dropped.
- Documented the required nginx header handling and added regression coverage. The final proxy library suite passed (624 tests, 4 ignored), along with the workspace check and repository hooks.

The trusted path still requires nginx to overwrite `X-Real-IP` and overwrite or append its observed client address as the final `X-Forwarded-For` entry. Thanks for raising the original reverse-proxy IP issue.

**Comment by dviejokfs** [comment]
@cymen I pushed `174134b` to close the two remaining review gaps. The proxy now loads the saved admin forwarding-header opt-in before it starts serving; if that first settings read fails, startup stops instead of temporarily treating visitor requests as loopback. I added regression tests for both startup paths and regenerated the web SDK from this PR's running backend.

I also exercised the current code in an isolated `/start-temps` slot. In Settings → Security, the switch saved successfully and stayed on after restart. With a synthetic blocked visitor IP in `X-Forwarded-For`, the live proxy returned 200 while the opt-in was off, then 403 after enabling it and again after restart. `cargo test --lib -p temps-proxy` passed (626 passed, 4 ignored), and `cargo check --lib` passed. The PR description has the commands and browser walkthrough.

File: crates/temps-proxy/src/server.rs [hunk 3]
```diff
@@ -32,6 +34,80 @@ use async_trait::async_trait;
...
+/// The initial IP policy must be known before Pingora accepts a request. A
+/// background refresh alone leaves a window where a persisted admin opt-in is
+/// ignored and loopback is used as the client IP for IP-based controls.
+async fn load_initial_forwarded_ip_trust(
+    config_service: &temps_config::ConfigService,
+    startup_enabled: bool,
+) -> Result<bool> {
+    let settings = config_service.get_settings().await.map_err(|error| {
+        anyhow::anyhow!("Cannot start proxy: failed to load forwarded-IP trust setting: {error}")
+    })?;
```
[the remark this answers is not in the document]
Comment (dviejokfs) [reply crates/temps-proxy/src/server.rs:60]: Fixed in 39747b0: the explicit startup override now sets trust to true without reading settings. The admin-setting path still loads before the proxy accepts traffic and aborts on an initial read error. Added a regression test for an unavailable settings read under the override; the full proxy suite passed (627 passed, 4 ignored).

File: crates/temps-cli/src/commands/proxy.rs [hunk 4]
```diff
@@ -179,6 +179,11 @@ pub struct ProxyCommand {
     /// Disable HTTP-to-HTTPS redirect (useful for local development without TLS)
     #[arg(long, env = "TEMPS_DISABLE_HTTPS_REDIRECT")]
     pub disable_https_redirect: bool,
+
+    /// Trust X-Forwarded-For or X-Real-IP from a same-host reverse proxy.
+    /// Enable only when that proxy overwrites or safely appends the client IP.
+    #[arg(long, env = "TEMPS_TRUST_LOOPBACK_FORWARDED_IP")]
```
[the remark this answers is not in the document]
Comment (dviejokfs) [reply crates/temps-cli/src/commands/proxy.rs]: Intentional exception for this PR: the operator explicitly requested a TEMPS_TRUST_LOOPBACK_FORWARDED_IP opt-in in addition to the database-backed admin switch. The environment flag is a process-wide startup override (especially useful for a standalone proxy), while the admin choice remains editable at runtime and trust remains off by default. Removing the env binding would remove requested behavior; the Security UI and PR description explain override precedence.

**Comment by dviejokfs** [comment]
@cymen One final follow-up on `5cb9577`: Greptile caught an override edge case, so the explicit env/CLI opt-in now starts with trust enabled even if the settings read fails; the saved admin setting is still loaded before traffic when there is no override. I added the regression test, and aligned the mock-listener test fixture with the equivalent fix now on `main` to clear the merge conflict. The proxy suite passed (628 passed, 4 ignored), `cargo check --lib` passed, and GitHub reports the PR mergeable. The new fork-head Actions jobs are still `action_required` pending maintainer approval. I replied inline to Greptile’s env-setting P2: the env opt-in is intentional alongside the admin UI, as requested for this PR.

**Comment by dviejokfs** [comment]
Pushed three follow-up commits addressing the outstanding review feedback:

- `23e09f9`: Removed the `TEMPS_TRUST_LOOPBACK_FORWARDED_IP` env var / `--trust-loopback-forwarded-ip` CLI flag entirely. On reflection this did violate the repository's "no env-var runtime config" rule as written — the setting already has a DB-backed column, admin UI, and loads before the proxy accepts traffic, so the bootstrap-before-DB-connection exception doesn't apply, and the override could silently keep trust on after an admin explicitly disabled it with no audit trail. The database-backed admin setting (Settings → Security → Proxy client IP) is now the sole control surface. Added a regression test for the refresh task's drop/exit behavior, which previously had no direct coverage.
- `f21e792`: Fixed a dual-stack-listener edge case where a same-host IPv4 connection arriving as `::ffff:127.0.0.1` wasn't recognized as loopback (`Ipv6Addr::is_loopback` only matches `::1`), leaving the feature silently inert even when correctly configured. Fail-safe either direction, so this is a correctness fix, not a security one.
- `cd1b0de`: Added a dedicated `FORWARDED_IP_TRUST_UPDATED` audit event (previous/new value) for this toggle, matching the existing pattern for other security-relevant settings in this file — `SETTINGS_UPDATED` alone doesn't carry field-level values.

`cargo test --lib -p temps-proxy -p temps-cli -p temps-config`: 1042 passed, 4 ignored. `cargo check --lib`, `cargo fmt --check`, and `cargo clippy --all-targets -- -D warnings` (run directly, not through the token-optimizing proxy some of my earlier runs used) are all clean on the affected crates. `bunx tsc --noEmit` and scoped ESLint pass on the updated `SecurityPage.tsx`.

File: crates/temps-proxy/src/server.rs [hunk 5]
```diff
@@ -32,6 +34,71 @@ use async_trait::async_trait;
...
+/// The initial IP policy must be known before Pingora accepts a request. A
+/// background refresh alone leaves a window where a persisted admin opt-in is
+/// ignored and loopback is used as the client IP for IP-based controls.
+async fn load_initial_forwarded_ip_trust(
+    config_service: &temps_config::ConfigService,
+) -> Result<bool> {
+    let settings = config_service.get_settings().await.map_err(|error| {
+        anyhow::anyhow!("Cannot start proxy: failed to load forwarded-IP trust setting: {error}")
+    })?;
+    Ok(settings.trust_loopback_forwarded_ip())
```
[the remark this answers is not in the document]
Comment (dviejokfs) [reply crates/temps-proxy/src/server.rs:61]: Fixed by updating the PR description rather than restoring the override: the CLI/env startup override was intentionally removed for the CLAUDE.md reasons discussed on the other thread (it duplicated the DB-backed setting and could silently override an admin's explicit choice with no audit trail). The description promising it was stale from before that removal — corrected now. `--trust-loopback-forwarded-ip` / `TEMPS_TRUST_LOOPBACK_FORWARDED_IP` are gone from both `temps proxy` and `temps serve`; the Settings → Security page is the sole control surface for every deployment topology, including a standalone proxy process.

**Comment by dviejokfs** [comment]
@greptile-apps review
